crates/hir-ty/src/infer/expr.rs

```diff
@@ -152,11 +152,20 @@ impl<'a> InferenceContext<'a> {
                 .1
             }
             Expr::TryBlock { body } => {
-                self.with_breakable_ctx(BreakableKind::Border, self.err_ty(), None, |this| {
-                    let _inner = this.infer_expr(*body, expected);
+                // The type that is returned from the try block
+                let try_ty = self.table.new_type_var();
+                if let Some(ty) = expected.only_has_type(&mut self.table) {
+                    self.unify(&try_ty, &ty);
+                }
+
+                // The ok-ish type that is expected from the last expression
+                let ok_ty = self.resolve_associated_type(try_ty.clone(), self.resolve_ops_try_ok());
+
+                self.with_breakable_ctx(BreakableKind::Block, ok_ty.clone(), None, |this| {
```

Comment:
I don't think that `BreakableKind::Block` is entirely correct, but as long as we do not check that `Expr::Try` is in a context where it can be used, this does not matter.

Reply:
Ye I imagine we'd need to add a new variant here once we check whether `Expr::Try` is in a correct context (we did for a week with an implementation I did, but that caused too many problems due to some things missing).